src/Interpreters/Context.cpp

```diff
@@ -1394,59 +1396,64 @@ StoragePtr Context::executeTableFunction(const ASTPtr & table_expression, const
         {
-            const auto & structure_hint = DatabaseCatalog::instance().getTable(getInsertionTable(), shared_from_this())->getInMemoryMetadataPtr()->getColumns();
-            bool use_columns_from_insert_query = true;
+            const auto & insert_structure = DatabaseCatalog::instance().getTable(getInsertionTable(), shared_from_this())->getInMemoryMetadataPtr()->getColumns();
+            DB::ColumnsDescription structure_hint;
 
-            /// use_structure_from_insertion_table_in_table_functions=2 means `auto`
-            if (select_query_hint && getSettingsRef().use_structure_from_insertion_table_in_table_functions == 2)
+            /// Insert table matches columns against SELECT expression by position, so we want to map
+            /// insert table columns to table function columns through names from SELECT expression.
+
+            auto insert_column = insert_structure.begin();
+            auto insert_structure_end = insert_structure.end();
+            auto virtual_column_names = table_function_ptr->getVirtualsToCheckBeforeUsingStructureHint();
+            bool asterisk = false;
+            const auto & expression_list = select_query_hint->select()->as<ASTExpressionList>()->children;
+            const auto * expression = expression_list.begin();
+
+            for (; expression != expression_list.end() && insert_column != insert_structure_end; ++expression)
             {
-                const auto * expression_list = select_query_hint->select()->as<ASTExpressionList>();
-                std::unordered_set<String> virtual_column_names = table_function_ptr->getVirtualsToCheckBeforeUsingStructureHint();
-                Names columns_names;
-                bool have_asterisk = false;
-                /// First, check if we have only identifiers, asterisk and literals in select expression,
-                /// and if no, we cannot use the structure from insertion table.
-                for (const auto & expression : expression_list->children)
+                if (auto * identifier = (*expression)->as<ASTIdentifier>())
                 {
-                    if (auto * identifier = expression->as<ASTIdentifier>())
+                    if (!virtual_column_names.contains(identifier->name()))
                     {
-                        columns_names.push_back(identifier->name());
-                    }
-                    else if (expression->as<ASTAsterisk>())
-                    {
-                        have_asterisk = true;
-                    }
-                    else if (!expression->as<ASTLiteral>())
-                    {
-                        use_columns_from_insert_query = false;
-                        break;
-                    }
-                }
+                        if (asterisk)
+                            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Asterisk cannot be mixed with column list in INSERT SELECT query.");
```

Comment:
Is it ok that we forbid it only when table function needs a structure hint?
So, this won't work:
```sql
insert into test select c1, * from file(data.parquet)
```
But this will work:
```sql
insert into test select c1, * from file(data.parquet, auto, 'c1 UInt32, c2 UInt32, ...')
```
Why don't set a hint and fallback to schema inference in this case?

Reply:
> the whole idea is to infer structure from insert table

Yes, but I think we should not throw an exception when we cannot do it and just don't set a hint. Most of the userrs don't know about `use_structure_from_insertion_table_in_table_functions` setting, so they will think that smth is broken when they cannot read data from file.

> or we are talking about using `use_structure_from_insertion_table_in_table_functions=auto` again

Maybe ... I just think that by default we definitely should use structure from insert table only when it's possible and don't throw an exception when it's not possible. We can do it under value `1` not `auto`. But if you think that  value `1` means "use structure from insertion table or throw", than yes, I am talking about `auto` again.